### ISSUE TITLE
Unable to create delta due to sig hashtable assertion error

### ISSUE THREAD

**Issue by jaknb** [librsync/librsync#115]
## Problem
I'm using librsync 2.0.1 and I'm facing the following error trying to create a delta:
```
rs_delta_begin: Assertion `!(sig)->hashtable || (sig)->hashtable->count == (sig)->count' failed.
```

## How to recreate
**PLEASE NOTE**: This will create a 1GB dummy file
``` bash
fallocate -l 1G big.dat
rdiff signature big.dat sigfile
rdiff delta sigfile big.dat deltafile
```
*Now you should encounter the problem.*

Clean the dummy files:
``` bash
rm big.dat sigfile deltafile
```

## System info (Ubuntu 16.04)
```
computer                  
    description: Computer
    product: VirtualBox
    vendor: innotek GmbH
    version: 1.2
    serial: [REMOVED]
    width: 64 bits
    capabilities: smbios-2.5 dmi-2.5 vsyscall32
    configuration: family=Virtual Machine uuid=[REMOVED]
  *-firmware
       description: BIOS
       vendor: innotek GmbH
       physical id: 0
       version: VirtualBox
       date: 12/01/2006
       size: 128KiB
       capabilities: isa pci cdboot bootselect int9keyboard int10video acpi
  *-memory
       description: System memory
       physical id: 1
       size: 3949MiB
  *-cpu
       product: Intel(R) Core(TM) i5-6600K CPU @ 3.50GHz
       vendor: Intel Corp.
       physical id: 2
       bus info: cpu@0
       width: 64 bits
       capabilities: fpu fpu_exception wp vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx rdtscp x86-64 constant_tsc rep_good nopl xtopology nonstop_tsc pni pclmulqdq ssse3 cx16 sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx rdrand hypervisor lahf_lm abm 3dnowprefetch rdseed clflushopt
  *-disk
       description: ATA Disk
       product: VBOX HARDDISK
       physical id: 0.0.0
       bus info: scsi@2:0.0.0
       logical name: /dev/sda
       version: 1.0
       serial: [REMOVED]
       size: 16GiB (17GB)
       capabilities: partitioned partitioned:dos
       configuration: ansiversion=5 logicalsectorsize=512 sectorsize=512 signature=f14c5211
```

**Comment by dbaarda** [comment]
Ugh! My bad. My pull that changed it to not add duplicate blocks meant that
this assertion is no longer correct when there are duplicate blocks. It
turns out none of our existing tests cover this case. I'll try to fix this
ASAP.

Note if you do a release build assertions are disabled and it should just
work, and will be faster.

**Comment by jaknb** [comment]
> Note if you do a release build assertions are disabled and it should just
work, and will be faster.

Oh, didn't know that. So to get the best performance I should stick to [the latest release](https://github.com/librsync/librsync/releases/tag/v2.0.0), right?

**Comment by dbaarda** [comment]
I didn't mean use an official release, I meant when building from the
latest git sources, in cmake change from a "DEBUG" build type to a
"RELEASE" build type. This sets NDEBUG true which disables all assert
checks. It also enables more agressive compiler optimisations, making it
run faster.

The current git head should be significantly faster than the last official
release, and we are just flushing things out like this bad assert statement
to make a new release soon.

Doing a RELEASE type build is a temporary workaround for the bug you found
before I can fix it.

On 28 Sep. 2017 6:40 pm, "jaknb" <notifications@github.com> wrote:

Note if you do a release build assertions are disabled and it should just
work, and will be faster.

Oh, didn't know that. So to get the best performance I should stick to the
latest release <https://github.com/librsync/librsync/releases/tag/v2.0.0>,
right?

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/librsync/librsync/issues/115#issuecomment-332769282>,
or mute the thread

**Comment by jaknb** [comment]
Ok. Thank you :)

Maybe [the docs](https://github.com/librsync/librsync/blob/master/doc/install.md) should be updated to contain the information about the different build types?

**Comment by dbaarda** [comment]
#57 is the bug asking for better docs. It needs some love; feel free to send a pull request :-)

I've just submitted a pull request #117 that fixes this bug and adds a test for it. Please let me know if that doesn't fix things for you.

**Comment by dbaarda** [comment]
I just merged #117. This is now fixed.
